core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -339,7 +341,7 @@ private[spark] class TaskSchedulerImpl(
           // We lost this entire executor, so remember that it's gone
           val execId = taskIdToExecutorId(tid)
 
-          if (executorIdToTaskCount.contains(execId)) {
+          if (executorIdToRunningTaskIds.contains(execId)) {
             reason = Some(
               SlaveLost(s"Task $tid was lost, so marking the executor as lost as well."))
             removeExecutor(execId, reason.get)
```

Comment:
Following the call to `removeExecutor` here, the `taskIdToTaskSetManager.get(tid)` call immediately following this `if` will be guaranteed to return `None`. This is going to slightly change the behavior of how `TaskState.LOST` events are handled, which, given my earlier comments, I believe is only relevant to Mesos fine-grained mode (I think this is underscored by the error message `s"Task $tid was lost, so marking the executor as lost as well."`, since this only makes sense if you assume one task per executor).

Given this, I'm planning to refactor this block slightly so that the body if the `if` corresponds to the old code and so that the code following the `if` is only run for non `TaskState.LOST` end reasons.